### PR TITLE
add deband shader

### PR DESCRIPTION
**Pull request by kato-megumi**
Based on Haasn's deband shader port from https://github.com/haasn/gentoo-conf/blob/xor/home/nand/.mpv/shaders/deband.glsl

### CODE REVIEW

File: Effects/Deband.hlsl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+// (Optional) Add some extra noise to the image. This significantly helps cover
+// up remaining banding and blocking artifacts, at comparatively little visual
+// quality. Higher = more grain. Setting it to 0 disables the effect.
+float grain;
+
+//!TEXTURE
+Texture2D INPUT;
+
+//!SAMPLER
+//!FILTER POINT
```
Comment (Blinue) [inline Effects/Deband.hlsl:45]: POINT or LINEAR? As far as I know, mpv shaders only supports linear sampling.

File: Effects/Deband.hlsl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+//!PARAMETER
+//!DEFAULT 4
+//!MIN 0
+// The number of debanding iterations to perform. Each iteration samples from
+// random positions, so increasing the number of iterations is likely to
+// increase the debanding quality. Conversely, it slows the shader down.
+// (Each iteration will use a multiple of the configured range, and a
+// successively lower THRESHOLD - so setting it much higher has little effect)
+float iterations;
```
Comment (Blinue) [inline Effects/Deband.hlsl:31]: Should be 'int'

File: Effects/Deband.hlsl [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+// Helper: Calculate a stochastic approximation of the avg color around a pixel
+float4 average(float2 pos, float range, inout float h)
+{
+    const float2 pt = GetInputPt();
+    // Compute a random rangle and distance
+    float dist = rand(h) * range;     h = permute(h);
+    float dir  = rand(h) * 6.2831853; h = permute(h);
+
+    float2 o = float2(cos(dir) * dist * pt.x, sin(dir) * dist * pt.y);
```
Comment (Blinue) [inline Effects/Deband.hlsl:66]:
``` hlsl
float2 o = float2(cos(dir), sin(dir)) * pt * dist;
```
This should be a little faster.

File: Effects/Deband.hlsl (hunk 1 above)
Comment (kato-megumi) [reply Effects/Deband.hlsl]: Yeah you're right. It should be linear.

File: Effects/Deband.hlsl [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+float4 Pass1(float2 pos)
+{
+    // Initialize the PRNG by hashing the position + a random uniform
+    float3 m = float3(pos, 4.6);
+    float h = permute(permute(permute(m.x)+m.y)+m.z);
+
+    // Sample the source pixel
+    float4 col = INPUT.SampleLevel(sam, pos, 0);
```
Comment (Blinue) [inline Effects/Deband.hlsl:87]: Because Magpie does not use the alpha channel, float3 is enough. Any subsequent float4 can also be changed to float3.

Comment (Blinue) [reply Effects/Deband.hlsl:87]: Point sampling is better here. Let's define two samplers.

**Comment by Blinue** [comment]
Comparison:
https://imgsli.com/MTUyNDQ5

Nice work!

**Review by Blinue — APPROVED** [review]

**Comment by Blinue** [comment]
LGTM! Thanks for contributing :)

**Pull request merged**